File: september/week1/partition_labels.py
from typing import List
from collections import OrderedDict

class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        hashdict = OrderedDict()
        for i,e in enumerate(S):
            if e not in hashdict:
                # start index, finish index
                hashdict[e] = [i,i]
            else:
                hashdict[e][1] = i

        for key,value in hashdict.items():
            s=value[0]
            f=value[1]
            for tar_key,tar_value in hashdict.items():
                if key==tar_key:
                    continue
                if s<=tar_value[0] and tar_value[1]<=f:
                    tar_value[0] = s
                    tar_value[1] = f
                if s<tar_value[0]<f or s<tar_value[1]<f:
                    s = tar_value[0] = min(s,tar_value[0])
                    f = tar_value[1] = max(f,tar_value[1])
                    
        ret = OrderedDict()
        for s,f in hashdict.values():
            if (s,f) not in ret:
                ret[(s,f)] = f-s+1

        ret = [val for val in ret.values()]
        return ret


if __name__ == "__main__":
    sol = Solution()
    print(sol.partitionLabels("ababcbacadefegdehijhklij"))

File: september/week1/PartitionLabels.cpp
#include <iostream>
#include <algorithm>
#include <cstdio>
#include <vector>
#include <string>

using namespace std;

int arr[26];
class Solution {
public:
	vector<int> partitionLabels(string S) {
		vector < int > idx;

		for (int i = 0; i < S.size(); i++) {
			int num = S[i] - 'a';
			arr[num] = i;
		}

		int n = 0;
		while (n < S.size()) {
			int end = arr[S[n] - 'a'];
			int val = end;
			for (int i = n; i < end; i++) {
				val = max(val, arr[S[i] - 'a']);
				end = val;
			}

			idx.push_back(val + 1);
			n = val + 1;
		}

		vector < int > ret;
		ret.push_back(idx[0]);
		for (int i = 1; i < idx.size(); i++)
			ret.push_back(idx[i] - idx[i - 1]);
		
		return ret;
	}
};

int main(void)
{
	Solution sol;

	vector < int > ans = sol.partitionLabels("ababcbacadefegdehijhklij");

	for (int i = 0; i < ans.size(); i++)
		cout << ans[i] << " ";
	return 0;
}